### ISSUE TITLE
[Bug] Winamp plugin conflicting with in_vgm for playback of ".vgm" files

### ISSUE THREAD

**Issue by the-muteKi** [vgmstream/vgmstream#1169]
When opening a .vgm file (i.e., timed chip instruction log files those found on sites like vgmrips.net) with the newest release of vgmstream in winamp, the in_vgmstream plugin is used to try to play it rather than in_vgm. While a workaround exists to rename .vgm files to .vgz (usually used to denote gzip-compressed vgm files), this seems like incorrect behavior based on the documentation of vgmstream, where renaming a file to '.vgmstream' (not merely '.vgm') will force use of the plugin; .vgm is not listed anywhere in the list of supported extensions.

**Comment by bnnm** [comment]
.vgm is one of vgmstream's extensions, but it should ignore regular .vgm (not recognized) unless incorrectly detected as another format.

Tried some random .vgm and seems correctly ignored, post some samples that show this issue. Also make sure you are using a recent version (see vgmstream.org/release page).

**Comment by the-muteKi** [comment]
This is with the most recent release (July 14, from commit 5aeecc7fd4da36507d0ea47fc2cafb1c0084833e). My Winamp version is 5.8 build 3660 from Oct. 2018.

**Comment by the-muteKi** [comment]
Here is an example file (zip compressed so that I can actually embed it in the comment):
[2B Chaos Emerald.zip](https://github.com/vgmstream/vgmstream/files/9126667/2B.Chaos.Emerald.zip)

**Comment by bnnm** [comment]
vgmstream ignores it on my end. What does it show in the info dialog?

A few validations related to misdetected .vgm were improved some months ago, so make sure you are putting the in_vgmstream dll in the correct place too.

**Comment by the-muteKi** [comment]
There's no popup that shows up at all. Is there anywhere the vgmstream DLL can go other than plugins, or are you referring to some of the support DLLs like xmp_vgmstream.dll (which are in the main winamp directory)

**Comment by bnnm** [comment]
DLL should go in plugins dir, but not sure if "new" winamps changed it, so just a general tip.

Anyway I was able to reproduce it, seems a mistake in how vgmstream rejects unsupported files. I'll try to fix it later.

**Comment by bnnm** [comment]
Should reject .vgm properly now so other plugins can play it.

**Comment by the-muteKi** [comment]
I'm sad to say that I've tried the newest release and it still doesn't seem to be working. Given that I've made sure to get the newest in_vgmstream and in_vgm releases, I'm a little surprised, Could the issue be tied to OS version? I'm using Windows 10; I recall file extension handling being different in 7, and have no idea about 11.

**Comment by bnnm** [comment]
I tested the new version from [here](https://vgmstream.org/downloads), with various options and seems ok (in_vgm plays it). Using windows 10 too. If you remove in_vgmstream does the .vgm play? I'm using Winamp 5.666 (last "original" one) but don't think it'd matter much.

**Comment by the-muteKi** [comment]
Yes, without vgmstream I don't have any conflicting issues. The file info tab option does show that with vgmstream installed that it is the plugin that is trying to handle. This may be an issue with my version of winamp being newer. I'll try 5.666 to see if that fixes anything; I don't think these newer versions of winamp, having used it for a while, introduce any features that make me prefer it over the previous ones.

And yes, I'm using the latest build of vgmstream from the website -- the version number in the info is g449bb5e0 (i.e., most recent commit)

**Comment by kode54** [comment]
Verdict: Winamp is broken ass shite.

**Comment by bnnm** [comment]
@the-muteKi try this debug build: [in_vgmstream.zip](https://github.com/vgmstream/vgmstream/files/9193132/in_vgmstream.zip)

Clear the playlist, close winamp, open it, then drag the .vgm. Then go to vgmstream's options (in preferences > plug-ins > input > vgmstream), press the "open log" button and report what it says.

Just to be clear, there are various possible behaviors here:
- vgmstream reports to winamp that it can play .vgm, but not actually play do it; so file is "hijacked" and added to the playlist without (other plugins can do the same, but less common)
- vgmstream does try to play the .vgm file as some misdetected format (for example as "FFmpeg supported" ) and is added with time. File info dialog should show something.
- Some other variation that may affect winamp 5.8?
The first case should've been fixed with the latest commits, report file info if you get the second. Maybe winamp 5.8 uses some of the weirder internal APIs and is triggering something else?

**Comment by the-muteKi** [comment]
I've switched back to 5.666 (since as you note that's the last 'official' version) but I've still been getting these issues. This is what the log says:

winampGetExtendedFileInfo_common: file [path to vgm] [this line repeats several times]
winamp_GetFileInfo:current
winamp_IsOurFile: hi.file://[path to vgm]
winamp_IsOurFile: invalid extension
winamp_getFileInfo: current

**Comment by the-muteKi** [comment]
In comparison for a vgz file (which in_vgm always picks up even if such a file is not actually compressed) this is what it looks like:

winamp_IsOurFile: hit.file://[path to vgz]
winamp_isOurFile: invalid extension

This check only occurs once [safe to presume before the file begins playing with in_vgm]. There is no call to winampGetExtendedFileInfo and there is never an appearance of the 
winamp_GetFileInfo: current 
line

**Comment by the-muteKi** [comment]
Trying this again with another file just to make sure that I'm not accidentally printing out invalid log entries, here's what I get when I drag a .vgm file into an empty playlist after having successfully played a .vgz file (and cleared it from the playlist). This clears up things and makes the log hopefully a bit easier to parse

winamp_IsOurFile: [path to VGM]
winamp_IsOurFile: invalid extension
winampGetExtendedFileInfo_common: file [path to VGM]
winampGetExtendedFileInfo_common: file [path to VGM]
winamp_GetFileInfo: file [path to VGM]

It seems like even though the extension is invalid, the plugin is still trying to get info from the file, and that's causing it not to send it to the other plugin to be played properly, if I'm understanding what's going on here.

**Comment by bnnm** [comment]
Try this version and report log please: [in_vgmstream.zip](https://github.com/vgmstream/vgmstream/files/9220806/in_vgmstream.zip)

In theory after winamp_IsOurFile should give up but there is a hidden "extended" API that may be affecting things. Not sure when it's used though, any less-common setting you have on? (media library, custom skins, replaygain, etc)

**Comment by the-muteKi** [comment]
![image](https://user-images.githubusercontent.com/109437174/182003189-64161e8f-fd39-4805-b4e5-198eed51c270.png)

**Comment by bnnm** [comment]
I think I get what's going on, which .vgm plugin are you using and how is the .dll for it and vgmstream named? (important). And how are they ordered in the input plugin list? (just in case).

Also check this and report log too please: [in_vgmstream.zip](https://github.com/vgmstream/vgmstream/files/9228900/in_vgmstream.zip)

**Comment by the-muteKi** [comment]
in_vgmstream.dll (unchanged from uploaded version) and in_vgmW.dll (not sure if this info is useful, but it shows as "VGM Input Plugin v.0.51.0" in the winamp input plugins list; downloaded from here: https://vgmrips.net/forum/viewtopic.php?t=112)

Here's the log:
![image](https://user-images.githubusercontent.com/109437174/182090026-a2357faf-6ce8-4bf0-9837-e1c660635eb3.png)

**Comment by bnnm** [comment]
Aha, if you rename in_vgmW.dll to in_vgm.dll, or in_vgmstream.dll to in_zvgmstream.dll (or anything, as long as vgmstream's .dll goes *after* in_vgm's dll in the plugin folder) that should fix the issue.

Here is what's going on:

Winamp first loads plugins as found in the plugin folder (ordered by filename), and asks every plugin if it supports the file (checking the contents). If nobody "claims" it, then it'll pass it to the *first* .dll that reports accepting the *extension* (without looking at the contents).

vgmstream now ignores non-streamed .vgm (used to "claim" it before), but in_vgm doesn't directly "claim" .vgm by content (it's common for most plugins to just leave it to the extension matching). Both .dll report accepting .vgm, so which plugin gets the .vgm is decided by the *filename order*.

This would be your order:
```
in_vgmstream.dll
in_vgmW.dll
```

But I had this, so the issue didn't trigger for me:
```
in_vgm.dll
in_vgmstream.dll
```

So basically renaming either would work as long as vgmstream goes after.

Can't really fix this, other than setting an option to not to report any extensions or something. That affects the open dialog and file types associations though.  For now will write some info in the guide. Maybe you can bring it up to the in_vgmW.dll's author to just call it in_vgm.dll, or "claim" the file during IsOurFile to avoid these clashes.

**Comment by the-muteKi** [comment]
OK. Thanks for the information. That was in fact the issue. I'll let the in_vgm developers know (in_vgmw is the unicode-supporting version of in_vgm) so that this issue doesn't persist with future releases, though since there's only need to install one version or the other, I can just keep renaming the file to in_vgm to fix this.

**Comment by the-muteKi** [comment]
I've let the in_vgm developers know about this issue. Since I don't think there's anything more I can reasonably expect from you all on this issue, I'll close the ticket. Thanks for the assistance in figuring out what was actually going on here.
